### PR TITLE
[11.x] Support eager loading with limit

### PR DESCRIPTION
**Pull request by staudenmeir**
As discussed: A new version of #26035 that integrates the [eloquent-eager-limit](https://github.com/staudenmeir/eloquent-eager-limit) package.

### Support for MySQL 5.7

According to the current [docs](https://laravel.com/docs/master/database#introduction), Laravel 11 won't support MySQL 5.7 anymore. I still included the legacy workaround for MySQL 5.7 since the necessary code already exists and MySQL 5.7 is still [being tested](https://github.com/laravel/framework/blob/master/.github/workflows/databases.yml#L11). The MySQL support in Laravel 11 is also still [being discussed](https://github.com/laravel/docs/pull/9113).

### CODE REVIEW

**Comment by staudenmeir** [comment]
The failing tests aren't related to this PR.

**Comment by decadence** [comment]
@staudenmeir Massive work 💪

Hope your other packages will have Laravel team attention. `HasManyDeep` is mind-blowing.

File: src/Illuminate/Database/Query/Grammars/MySqlGrammar.php [hunk 1]
```diff
@@ -94,6 +95,82 @@ protected function compileIndexHint(Builder $query, $indexHint)
...
+
+    /**
+     * Determine whether to use a legacy group limit clause for MySQL < 8.0.
+     *
+     * @param  \Illuminate\Database\Query\Builder  $query
+     * @return bool
+     */
+    public function useLegacyGroupLimit(Builder $query)
+    {
+        $version = $query->getConnection()->getReadPdo()->getAttribute(PDO::ATTR_SERVER_VERSION);
```
Comment (hafezdivandari) [inline src/Illuminate/Database/Query/Grammars/MySqlGrammar.php:119]:
```suggestion
        $version = $query->getConnection()->getServerVersion();
```

File: src/Illuminate/Database/Query/Grammars/SQLiteGrammar.php [hunk 2]
```diff
@@ -184,6 +185,25 @@ protected function compileJsonContainsKey($column)
...
 
+    /**
+     * Compile a group limit clause.
+     *
+     * @param  \Illuminate\Database\Query\Builder  $query
+     * @return string
+     */
+    protected function compileGroupLimit(Builder $query)
+    {
+        $version = $query->getConnection()->getReadPdo()->getAttribute(PDO::ATTR_SERVER_VERSION);
```
Comment (hafezdivandari) [inline src/Illuminate/Database/Query/Grammars/SQLiteGrammar.php:196]:
```suggestion
        $version = $query->getConnection()->getServerVersion();
```

**Comment by taylorotwell** [comment]
Thanks @staudenmeir! 🥳

File: src/Illuminate/Database/Query/Grammars/Grammar.php [hunk 3]
```diff
@@ -60,6 +60,17 @@ public function compileSelect(Builder $query)
             return $this->compileUnionAggregate($query);
         }
 
+        // If a "group limit" is in place, we will need to compile the SQL to use a
```
Comment (VinceSanityyy) [inline src/Illuminate/Database/Query/Grammars/Grammar.php:63]: a

File: src/Illuminate/Database/Query/Grammars/MySqlGrammar.php (hunk 1 above)
Comment (hafezdivandari) [reply src/Illuminate/Database/Query/Grammars/MySqlGrammar.php:119]: Done: #49723

**Comment by thanhnguyenphuocnhat** [comment]
Great work!!!

**Comment by kon3ko** [comment]
Thank a lot <3

**Comment by mokhosh** [comment]
can't wait for your other eloquent extension packages to be included in the core.

**Comment by lmottasin** [comment]
Thank you.

File: src/Illuminate/Database/Query/Grammars/MySqlGrammar.php [hunk 4]
```diff
@@ -94,6 +95,82 @@ protected function compileIndexHint(Builder $query, $indexHint)
...
+
+        if (isset($offset)) {
+            $offset = (int) $offset;
+            $limit += $offset;
+
+            $query->offset = null;
+        }
+
+        $column = last(explode('.', $query->groupLimit['column']));
+        $column = $this->wrap($column);
```
Comment (parallels999) [inline src/Illuminate/Database/Query/Grammars/MySqlGrammar.php:145]: @staudenmeir hi. why remove the table's name?

```diff
-$column = last(explode('.', $query->groupLimit['column']));
-$column = $this->wrap($column);
+$column = $this->wrap($query->groupLimit['column']);
```

Look at this:
`Column not found: 1054 Unknown column 'auditable_id' in 'field list'`
`https://github.com/owen-it/laravel-auditing/issues/974`

is that related?

Comment (staudenmeir) [reply src/Illuminate/Database/Query/Grammars/MySqlGrammar.php:145]: Removing the table name is not the issue here. If the column isn't found without the table name, it can't be found it with the table name either.

**Comment by pintend** [comment]
I am updating an old Laravel project (originally 8) and this is an undocumented breaking change for Laravel 10 to 11. The project had a query which was ordering by pivot columns which are no longer pivot_* (need to now just do table.column)

so i guess this is just a psa

https://github.com/laravel/docs/pull/10507
